fix: recognize dollar sign in identifier

ardislu opened the pull request:
## Description

This PR updates `errorSignatureRegex`, `eventSignatureRegex`, `functionSignatureRegex`, and `structSignatureRegex` to allow the dollar sign (`$`) to be used in an identifier. It also disallows identifiers that begin with a number as the first symbol. Reference: [Solidity docs](https://docs.soliditylang.org/en/latest/grammar.html#a4.SolidityLexer.Identifier).

I also updated the regexr links for each regex.

Fixes #163.

## Additional Information

- [X] I read the [contributing guide](https://github.com/wagmi-dev/abitype/blob/main/.github/CONTRIBUTING.md)
- [ ] I added documentation related to the changes made.
- [X] I added or updated tests related to the changes made.

Your ENS/address: `ardis.eth`

Raiden1411 commented:
Could you also change the regexes of `abiParameterWithTupleRegex` and `abiParameterWithoutTupleRegex`?

And could you also change the the type `ValidCharacters` to support the `$`

ardislu commented:
@Raiden1411

I updated the `<name>` capture group in `abiParameterWithoutTupleRegex` and `abiParameterWithTupleRegex` to recognize the `$` character so these two are now aligned with the others.

I think the `<type>` capture group in these regexes may also need to be updated, because it looks like [user-defined types can also be named with any valid identifier](https://docs.soliditylang.org/en/latest/grammar.html#a4.SolidityParser.typeName), but I'm not sure how to go about updating the regex for the types.

I also updated the `ValidCharacters` type to add `$`.

Raiden1411 commented:
You can just update the `<type>` capture group on `abiParameterWithoutTupleRegex` with the new pattern. The tuple regex `<type>` capture group when extracted is used to recursively call `parseAbiParameter` which will then use `abiParameterWithoutTupleRegex` if the splited parameter is not a tuple. You can add a unit test to ensure the correctness of this behavior.

Raiden1411 commented:
Also, if you could add a changeset to this PR that would be appreciated.

ardislu commented:
@Raiden1411 

Got it, makes sense.

I updated the `<type>` regex in `abiParameterWithoutTupleRegex`. I also noticed `typeWithoutTupleRegex` (within `structs.ts`) uses the same regex so I updated that as well.

The existing unit tests look like they already test for the recursive behavior, so I didn't add any new ones. The updated regexes pass all existing unit tests.

I added a changeset. Not sure if I did it correctly, let me know if I need to make any changes.

Thanks for your help!

Raiden1411 commented:
Thank you, nice check! I'd say even though the regexes pass the current tests it would be good to have one or two dedicated tests to check that an identifier cannot start with a number, can contain the `$` char and it throws accordingly. After that this should be good to merge.

ardislu commented:
@Raiden1411 

I added unit tests to confirm:

- `$`, `$_a9`, `_`, and `abc$_9` are all VALID identifier names
- `9abc` is an INVALID identifier name

I think my tests should cover all places where I updated the regex, but do let me know if I should add tests in any other files.

Thanks!

tmm commented on .changeset/dull-nails-move.md:
```diff
@@ -0,0 +1,6 @@
+"abitype": patch
+---
+
+Allowed "$" char in identifiers on both type level and runtime.
+Disallows number strings and identifiers that start with a number.
```
```suggestion
Fixed human-readable identifiers to allow single leading dollar-sign or underscore, and disallow single leading number.
```

tmm reviewed (commented):
nice work!
